edx/edx-ora2
# Submission lock API

**TL;DR -** Add an API for CRUD actions on submission locks.

JIRA: [AU-54](https://openedx.atlassian.net/browse/AU-54)
Linked PR: https://github.com/edx/edx-platform/pull/28816

**What changed?**

- Added new `SubmissionGradingLock ` model to support locking a submission
- Added `StaffGraderMixin` to add XBlock JSON handler methods:
    - `check_submission_lock`
    - `claim_submission_lock`
    - `delete_submission_lock`

**API Documentation**
All endpoints follow the structure (due to them being `XBlock.json_handler`s)

```
POST http://{{lms_url}}/courses/{{course_id}}/xblock/{{block-locator}}/handler/{{name}}

BODY {
    "submission_id": <submission_uuid>
}

RESPONSE {
    "submission_uuid": <submisison_uuid>,
    "owner_id": <lock_owner_anon_id>,
    "created_at": <lock_creation_timestamp>
}

ERRORS
- 400 - missing submission_id
- 403 - lock contested OR insufficient permissions
- 405 - not logged in
```

**Developer Checklist**


## Testing Instructions

**Test Setup**

The best way to test is using a tool like Postman to exercise the APIs. First, since we're plugged in to LMS CSRF and Auth framework:

1. Capture login session (using Postman/similar):
    1. Use cookie synchronization to copy over cookies from an active login session.
    2. Copy `csrftoken` value from cookie, add it to a new `X-CSRFToken` header.
2. Find a submission ID to work with for your course (for example, by looking at [Django Admin > Submissions](http://localhost:18000/admin/submissions/submission/)

Verify the following cases:

**Access tests** - These are built-in by `XBlock.json_handler` but might be good to verify anyways.
1. A non-logged in user gets a `405` (Unauthenticated) for each request.
2. A non-staff user for the given course receives a `403` (Forbidden) for each request.

**Method tests**
As a user with staff permissions:
1. `get_submission_lock` returns submission lock info or Empty `{}`
2. `claim_submission_lock` for a submission without a lock creates a new lock and returns lock info.
3. `clear_submission_lock` for a submission with (or without) a lock clears the lock and returns `{}`

Next, we want to verify only the owner of current locks can renew/clear them. As (probably a different) user with staff permissions:
1. `get_submission_lock` returns submission lock info on the other user's lock.
2. `claim_submission_lock` for a lock owned by another user returns a `403` with response `{ "Error": "Submission already locked"}`
2. `clear_submission_lock` for a lock owned by another user returns a `403` with response `{ "Error": "Unable to clear submission lock"}`

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:


FYI: @edx/masters-devs-gta
